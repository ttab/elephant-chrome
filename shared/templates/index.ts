import type { Block } from '@ttab/elephant-api/newsdoc'

export { articleDocumentTemplate as article } from './articleDocumentTemplate.js'
export { flashDocumentTemplate as flash } from './flashDocumentTemplate.js'
export { factboxDocumentTemplate as factbox } from './factboxDocumentTemplate.js'
export { assignmentPlanningTemplate as assignment } from './assignmentPlanningTemplate.js'
export { planningDocumentTemplate as planning } from './planningDocumentTemplate.js'
export { eventDocumentTemplate as event } from './eventDocumentTemplate.js'
export { editorialInfoDocumentTemplate as editorialInfo } from './editorialInfoDocumentTemplate.js'

export interface TemplatePayload {
  title?: string
  meta?: {
    'tt/slugline'?: Block[]
    'core/newsvalue'?: Block[]
    'core/planning-item'?: Block[]
    'core/description'?: Block[]
    'core/event'?: Block[]
  }
  links?: {
    'core/section'?: Block[]
    'core/story'?: Block[]
    'tt/wire'?: Block[]
    'core/event'?: Block[]
  }
}
