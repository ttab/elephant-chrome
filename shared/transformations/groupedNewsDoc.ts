import { type GetDocumentResponse } from '@ttab/elephant-api/repository'
import { Block, type Document } from '@ttab/elephant-api/newsdoc'

import type {
  EleDocumentResponse,
  EleDocument,
  EleBlockGroup
} from '@/shared/types/index.js'

import { newsDocToSlate, slateToNewsDoc } from './newsdoc/index.js'


/**
 *  Convert repository format NewsDoc to grouped YDocument format
 */
export const toGroupedNewsDoc = (payload: GetDocumentResponse): EleDocumentResponse => {
  // Create clone of document so not to change original document/cause sideffects
  const { document, version, isMetaDocument, mainDocument } = structuredClone(payload)

  if (!document) {
    throw new Error('GetDocumentResponse contains no document')
  }

  if (document.type === 'core/planning-item') {
    assertPlanningHasNecessaryProperties(document)
  }

  const yDocument: EleDocument = {
    ...document,
    content: newsDocToSlate(document.content, document.type === 'tt/print-article'),
    meta: group(document.meta || [], 'type'),
    links: group(document.links || [], 'type')
  }

  return {
    version: version.toString(),
    isMetaDocument,
    mainDocument,
    document: yDocument
  }
}


/**
 *  Convert grouped YDocument format to the repository format NewsDoc
 */
export const fromGroupedNewsDoc = (payload: EleDocumentResponse): { document: Document } & Omit<GetDocumentResponse, 'document'> => {
  const { document, version, isMetaDocument, mainDocument } = payload

  if (!document) {
    throw new Error('YDocumentResponse contains no document')
  }

  const newsDocument = {
    ...document,
    content: slateToNewsDoc(document.content) || [],
    meta: ungroup(document.meta),
    links: ungroup(document.links)
  }

  if (document.type === 'core/planning-item') {
    assertPlanningHasNoEmptyProperties(newsDocument)
  }

  return {
    version: BigInt(version),
    document: newsDocument,
    isMetaDocument,
    mainDocument
  }
}


/**
 * Group Blocks with same group key in arrays
 *
 * @param Block[]
 * @param groupKey
 * @returns YBlockGroup
 */
export function group(objects: Block[], groupKey: keyof Block): EleBlockGroup {
  const groupedObjects: EleBlockGroup = {}

  objects.forEach((object) => {
    let key = object[groupKey] as string | undefined

    // Fallback to underscore if no groupKey is found
    if (!key) {
      key = '_'
    }

    if (!groupedObjects[key]) {
      groupedObjects[key] = []
    }

    const newObj = { ...object }

    Object.entries(newObj).forEach(([key, value]) => {
      if (Array.isArray(value)) {
        // @ts-expect-error Can't get this ts error to go away
        newObj[key] = group(value, groupKey)
      }
    })

    // @ts-expect-error Can't get this ts error to go away
    groupedObjects[key].push(newObj)
  })

  return groupedObjects
}


/**
 * Reverts the operation of the group function
 *
 * @param YBlockGroup
 * @returns Block[]
 */
export function ungroup(obj: EleBlockGroup): Block[] {
  const result: Block[] = []

  Object.keys(obj).forEach((key) => {
    if (Array.isArray(obj[key])) {
      obj[key].forEach((item) => {
        // Ignore __inProgress items
        if (!item.__inProgress) {
          const newObj = Block.create({
            ...item,
            meta: ungroup(item.meta as unknown as EleBlockGroup || {}),
            links: ungroup(item.links as unknown as EleBlockGroup || {}),
            content: ungroup(item.content as unknown as EleBlockGroup || {})
          })

          result.push(newObj as unknown as Block)
        }
      })
    }
  })

  return result
}

/**
 *  Assert that we don't leave empty sluglines or descriptions on planning documents
 */
function assertPlanningHasNoEmptyProperties(obj: Document | Block): void {
  obj.meta = obj.meta.filter((block) => {
    if (block.type === 'core/description' && !block.data.text) {
      return false
    } else if (block.type === 'tt/slugline' && !block.value) {
      return false
    }

    return true
  })

  obj.meta.forEach((block) => {
    if (block.type === 'core/assignment') {
      assertPlanningHasNoEmptyProperties(block)
    }
  })
}


/**
 *  Assert that slugline and both internal and public descriptions exist on planning documents
 */
function assertPlanningHasNecessaryProperties(obj: Document | Block): void {
  const internalDesc = obj.meta.find((item) => item.type === 'core/description' && item.role === 'internal')
  const publicDesc = obj.meta.find((item) => item.type === 'core/description' && item.role === 'public')
  const slugline = obj.meta.find((item) => item.type === 'tt/slugline')

  if (!slugline) {
    obj.meta.push(Block.create({
      type: 'tt/slugline'
    }))
  }

  if (!internalDesc) {
    obj.meta.push(Block.create({
      type: 'core/description',
      role: 'internal',
      data: { text: '' }
    }))
  }

  if (!publicDesc) {
    obj.meta.push(Block.create({
      type: 'core/description',
      role: 'public',
      data: { text: '' }
    }))
  }

  obj.meta.forEach((block) => {
    if (block.type === 'core/assignment') {
      assertPlanningHasNecessaryProperties(block)
    }
  })
}
