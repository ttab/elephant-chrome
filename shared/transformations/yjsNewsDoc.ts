import * as Y from 'yjs'
import { type Document } from '@hocuspocus/server'

import type {
  EleBlock,
  EleDocumentResponse
} from '@/shared/types/index.js'
import { toYMap } from './toYMap.js'
import { slateNodesToInsertDelta, yTextToSlateElement } from '@slate-yjs/core'
import createHash from '../createHash.js'
import { type TBElement } from '@ttab/textbit'
import { transformYXmlTextNodes } from './lib/transformYXmlTextNodes.js'

/**
 * Convert a grouped YDocumentResponse a yjs structure and add it to the provided Y.Doc
 */
export function toYjsNewsDoc(eleDoc: EleDocumentResponse, yDoc: Document | Y.Doc): void {
  const yMap = yDoc.getMap('ele')
  const { document, version } = eleDoc

  if (!document) {
    throw new Error('YDocumentResponse contains no document')
  }

  const { meta, links, content, ...properties } = document
  yMap.set('meta', toYMap(meta))
  yMap.set('links', toYMap(links))
  yMap.set('root', toYMap(properties, new Y.Map()))

  // Slate text to yjs
  const yContent = new Y.XmlText()
  yContent.applyDelta(
    slateNodesToInsertDelta(content)
  )

  yMap.set('content', yContent)

  // Set version
  const yVersion = yDoc.getMap('version')
  yVersion?.set('version', version)

  // Create original hash based on eleDocument content and store in yjs structure
  const originalHash = createHash(JSON.stringify(eleDoc.document))
  const yOriginalHash = yDoc.getMap('hash')
  yOriginalHash?.set('hash', originalHash)
}


/**
 * Convert a yjs structure to a grouped YDocumentResponse
 */
export function fromYjsNewsDoc(yDoc: Y.Doc): {
  documentResponse: EleDocumentResponse
  updatedHash: number | undefined
  originalHash: number
} {
  const yMap = yDoc.getMap('ele')

  const root = yMap.get('root') as Y.Map<unknown>
  const { uuid, type, uri, url, title, language } = root.toJSON() as Record<string, string>

  const meta = transformYXmlTextNodes(yMap.get('meta') as Y.Map<unknown>) as Record<string, EleBlock[]>

  const links = (yMap.get('links') as Y.Map<unknown>).toJSON() || {}

  const yContent = yMap.get('content') as Y.XmlText
  const content = yContent.length ? yTextToSlateElement(yContent).children : []

  const responseDocument = {
    version: yDoc.getMap('version').get('version') as string,
    isMetaDocument: false,
    mainDocument: '',
    document: {
      uuid,
      type,
      uri,
      url,
      title,
      content: content as TBElement[],
      meta,
      links,
      language
    }
  }

  const currentHash = createHash(JSON.stringify(responseDocument.document))
  const originalHash = (yDoc.getMap('hash')?.get('hash') || 0) as number

  // Return response, original hash from yjs structure and newly calculated hash
  return {
    documentResponse: responseDocument,
    updatedHash: (currentHash !== originalHash) ? currentHash : undefined,
    originalHash
  }
}
