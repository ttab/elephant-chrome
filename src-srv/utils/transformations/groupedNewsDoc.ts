import {
  Block,
  type Document,
  type GetDocumentResponse
} from '@/shared/protos/service.js'

import type {
  YDocumentResponse,
  YDocument,
  YBlockGroup
} from '@/shared/types/index.js'

import { newsDocToSlate, slateToNewsDoc } from './newsdoc/index.js'


/**
 *  Convert repository format NewsDoc to grouped YDocument format
 */
export const toGroupedNewsDoc = (payload: GetDocumentResponse): YDocumentResponse => {
  const { document, version } = payload

  if (!document) {
    throw new Error('GetDocumentResponse contains no document')
  }

  if (document.type === 'core/planning') {
    assertPlanningHasNecessaryProperties(document)
  }

  // const { meta, links, content, ...properties } = document
  const yDocument: YDocument = {
    ...document,
    content: newsDocToSlate(document.content),
    meta: group(document.meta || [], 'type'),
    links: group(document.links || [], 'type')
  }

  return {
    version: version.toString(),
    document: yDocument
  }
}


/**
 *  Convert grouped YDocument format to the repository format NewsDoc
 */
export const fromGroupedNewsDoc = async (payload: YDocumentResponse): Promise<GetDocumentResponse> => {
  const { document, version } = payload

  if (!document) {
    throw new Error('YDocumentResponse contains no document')
  }

  const newsDocument = {
    ...document,
    // FIXME: Why is this async, really? This forces this func to be async!
    content: await slateToNewsDoc(document.content) || [],
    meta: ungroup(document.meta),
    links: ungroup(document.links)
  }

  assertPlanningHasNoEmptyProperties(newsDocument)

  return {
    version: BigInt(version),
    document: newsDocument
  }
}


/**
 * Group Blocks with same group key in arrays
 *
 * @param Block[]
 * @param groupKey
 * @returns YBlockGroup
 */
function group(objects: Block[], groupKey: keyof Block): YBlockGroup {
  const groupedObjects: YBlockGroup = {}

  objects.forEach(object => {
    const key = object[groupKey] as string | undefined
    if (!key) return

    if (!groupedObjects[key]) {
      groupedObjects[key] = []
    }

    const newObj = { ...object }

    Object.entries(newObj).forEach(([key, value]) => {
      if (Array.isArray(value)) {
        // @ts-expect-error Can't get this ts error to go away
        newObj[key] = group(value, groupKey)
      }
    })

    // @ts-expect-error Can't get this ts error to go away
    groupedObjects[key].push(newObj)
  })

  return groupedObjects
}


/**
 * Reverts the operation of the group function
 *
 * @param YBlockGroup
 * @returns Block[]
 */
function ungroup(obj: YBlockGroup): Block[] {
  const result: Block[] = []

  Object.keys(obj).forEach(key => {
    if (Array.isArray(obj[key])) {
      obj[key].forEach((item) => {
        // Ignore __inProgress items
        if (!item.__inProgress) {
          const newObj = Block.create({
            ...item,
            meta: ungroup(item.meta as unknown as YBlockGroup || {}),
            links: ungroup(item.links as unknown as YBlockGroup || {}),
            content: ungroup(item.content as unknown as YBlockGroup || {})
          })

          result.push(newObj as unknown as Block)
        }
      })
    }
  })

  return result
}

/**
 *  Assert that we don't leave empty sluglines or descriptions on planning documents
 */
function assertPlanningHasNoEmptyProperties(obj: Document | Block): void {
  obj.meta = obj.meta.filter(block => {
    if (block.type === 'core/description' && !block.data.text) {
      return false
    } else if (block.type === 'tt/slugline' && !block.value) {
      return false
    }

    return true
  })

  obj.meta.forEach(block => {
    if (block.type === 'core/assignment') {
      assertPlanningHasNoEmptyProperties(block)
    }
  })
}


/**
 *  Assert that slugline and both internal and public descriptions exist on planning documents
 */
function assertPlanningHasNecessaryProperties(obj: Document | Block): void {
  const internalDesc = obj.meta.find(item => item.type === 'core/description' && item.role === 'internal')
  const publicDesc = obj.meta.find(item => item.type === 'core/description' && item.role === 'public')
  const slugline = obj.meta.find(item => item.type === 'tt/slugline')

  if (!slugline) {
    obj.meta.push(Block.create({
      type: 'tt/slugline'
    }))
  }

  if (!internalDesc) {
    obj.meta.push(Block.create({
      type: 'core/description',
      data: { text: '' }
    }))
  }

  if (!publicDesc) {
    obj.meta.push(Block.create({
      type: 'core/description',
      data: { text: '' }
    }))
  }

  obj.meta.forEach(block => {
    if (block.type === 'core/assignment') {
      assertPlanningHasNecessaryProperties(block)
    }
  })
}
