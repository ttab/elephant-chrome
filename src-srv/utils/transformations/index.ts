export {
  slateToNewsDoc,
  newsDocToSlate
} from './newsdoc/index.js'
