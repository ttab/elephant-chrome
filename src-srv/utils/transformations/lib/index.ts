export { toYMap } from './toYMap.js'
export { toSlateYXmlText } from './toSlateYXmlText.js'
export { textToNewsDoc } from './textToNewsdoc.js'
