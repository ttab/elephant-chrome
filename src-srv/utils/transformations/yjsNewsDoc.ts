import * as Y from 'yjs'
import { type Document } from '@hocuspocus/server'

import type {
  EleDocumentResponse
} from '@/shared/types/index.js'
import { toYMap } from './lib/toYMap.js'
import { slateNodesToInsertDelta, yTextToSlateElement } from '@slate-yjs/core'
import createHash from '@/shared/createHash.js'


/**
 * Convert a grouped YDocumentResponse a yjs structure and add it to the provided Y.Doc
 */
export function toYjsNewsDoc(eleDoc: EleDocumentResponse, yDoc: Document | Y.Doc): void {
  const yMap = yDoc.getMap('ele')
  const { document, version } = eleDoc

  if (!document) {
    throw new Error('YDocumentResponse contains no document')
  }

  const { meta, links, content, ...properties } = document
  yMap.set('meta', toYMap(meta))
  yMap.set('links', toYMap(links))
  yMap.set('root', toYMap(properties, new Y.Map()))

  // Slate text to yjs
  const yContent = new Y.XmlText()
  yContent.applyDelta(
    slateNodesToInsertDelta(content)
  )
  yMap.set('content', yContent)

  // Set version
  const yVersion = yDoc.getMap('version')
  yVersion?.set('version', version)

  // Create original hash based on eleDocument content and store in yjs structure
  const originalHash = createHash(JSON.stringify(eleDoc.document))
  const yOriginalHash = yDoc.getMap('hash')
  yOriginalHash?.set('hash', originalHash)
}


/**
 * Convert a yjs structure to a grouped YDocumentResponse
 */
export async function fromYjsNewsDoc(yDoc: Y.Doc): Promise<{
  documentResponse: EleDocumentResponse
  updatedHash: number | undefined
  originalHash: number
}> {
  const yMap = yDoc.getMap('ele')

  const root = yMap.get('root') as Y.Map<unknown>
  const { uuid, type, uri, url, title, language } = root.toJSON()

  const meta = (yMap.get('meta') as Y.Map<unknown>).toJSON() || {}
  const links = (yMap.get('links') as Y.Map<unknown>).toJSON() || {}

  const yContent = yMap.get('content') as Y.XmlText
  const content = yContent.toString() ? await yTextToSlateElement(yContent).children : []

  const responseDocument = {
    version: yDoc.getMap('version').get('version') as string,
    document: {
      uuid,
      type,
      uri,
      url,
      title,
      content,
      meta,
      links,
      language
    }
  }

  const currentHash = createHash(JSON.stringify(responseDocument.document))
  const originalHash = (yDoc.getMap('hash')?.get('hash') || 0) as number

  // Return response, original hash from yjs structure and newly calculated hash
  return {
    documentResponse: responseDocument,
    updatedHash: (currentHash !== originalHash) ? currentHash : undefined,
    originalHash
  }
}
