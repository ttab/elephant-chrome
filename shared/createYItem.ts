import * as Y from 'yjs'
import { Block, type Document } from '@ttab/elephant-api/newsdoc'
import type { Wire } from '@/shared/schemas/wire.js'
import { assignmentPlanningTemplate } from '@/shared/templates/assignmentPlanningTemplate.js'
import type { DeliverableType } from '@/shared/templates/lib/getDeliverableType.js'
import { group, toGroupedNewsDoc } from '@/shared/transformations/groupedNewsDoc.js'
import { toYMap } from '@/shared/transformations/toYMap.js'
import { toYjsNewsDoc } from '@/shared/transformations/yjsNewsDoc.js'
import type { IDBAuthor } from '../src/datastore/types.js'

/**
* General function to create a new document as Y.Doc from a template
* @returns [string, Y.Doc]
*/
interface CreateDocumentParams<T> {
  template: (documentId: string, payload?: T) => Document
  inProgress?: boolean
  payload?: T
  createdDocumentIdRef?: React.MutableRefObject<string | undefined>
  documentId?: string
}

export function createDocument<T>({
  template,
  inProgress,
  payload,
  createdDocumentIdRef,
  documentId
}: CreateDocumentParams<T>): [string, Y.Doc] {
  // Use provided documentId or generate a new one
  const docId = documentId || crypto.randomUUID()

  // Set generated documentId to ref so that it can be
  // accessed from creating component
  if (createdDocumentIdRef) {
    createdDocumentIdRef.current = docId
  }

  const yDoc = new Y.Doc()

  toYjsNewsDoc(
    toGroupedNewsDoc({
      version: 0n,
      isMetaDocument: false,
      mainDocument: '',
      document: template(docId, payload)
    }),
    yDoc
  )

  if (inProgress) {
    yDoc.getMap('__meta').set('isInProgress', true)
  }

  return [docId, yDoc]
}

/**
* Specific function to create a new assignment as Y.Map from a template
* and append it to the meta yMap
* @returns void
*/
export function appendAssignment({
  assignee,
  document,
  type,
  inProgress,
  slugLine,
  title,
  wire,
  assignmentData
}: {
  document: Y.Doc
  assignee?: IDBAuthor | null | undefined
  type: 'text' | 'flash' | 'graphic' | 'picture' | 'video' | 'picture/video'
  inProgress?: { sub: string }
  slugLine?: string
  title?: string
  wire?: Wire
  assignmentData?: Block['data']
}): [number, Y.Map<unknown>] {
  const meta = document.getMap('ele').get('meta') as Y.Map<unknown>

  // Get slugline from planning
  const slugLineArray = meta?.get('tt/slugline') as Y.Array<unknown>
  const slugLineYXml = slugLineArray?.get(0) as Y.Map<unknown>
  const slugLineFromPlanning = (slugLineYXml?.get('value') as Y.XmlText)?.toString() as string || undefined

  // Check if 'core/assignment' exists
  if (!meta.has('core/assignment')) {
    meta.set('core/assignment', new Y.Array())
  }

  // Get existing assignments
  const yAssignments = meta.get('core/assignment') as Y.Array<unknown>

  const yMeta = meta.get('core/planning-item') as Y.Array<Y.Map<unknown>>
  const planningDate = (yMeta?.get(0)?.get('data') as Y.Map<unknown>)
    ?.get('start_date') as string

  // Create new assignment from template
  const assignment = assignmentPlanningTemplate({
    assignee,
    assignmentType: type,
    planningDate,
    slugLine: type !== 'flash' ? slugLine || slugLineFromPlanning : undefined,
    title: title,
    wire,
    assignmentData
  })

  // Append __inProgress if needed
  if (inProgress) {
    // @ts-expect-error We need to override Block to add this property
    assignment.__inProgress = inProgress
  }

  // Group assignment
  const [groupedAssignment] = group([assignment], 'type')['core/assignment']

  // Convert to YMap
  const yAssignment = toYMap(
    groupedAssignment as unknown as Record<string, unknown>,
    new Y.Map()
  )

  // Push to existing assignments
  yAssignments.push([yAssignment])

  return [
    yAssignments.length - 1,
    yAssignment
  ]
}

/**
* Specific function to create a new article in Y.Doc as Y.Map from a template
*/
export function appendDocumentToAssignment({ document, id, index, slug, type }: {
  document: Y.Doc
  id: string
  index: number
  slug?: string
  type: DeliverableType
}): void {
  // Get meta yMap
  const meta = document.getMap('ele').get('meta') as Y.Map<unknown>

  // Get assignmentLinks
  const assignmentLinks = ((meta
    .get('core/assignment') as Y.Array<unknown>)
    .get(index) as Y.Map<unknown>)
    .get('links') as Y.Map<unknown>

  // Check if deliverableType exists
  if (!assignmentLinks.has(`core/${type}`)) {
    assignmentLinks.set(`core/${type}`, new Y.Array())
  }
  // Get existing articles
  const yDeliverables = assignmentLinks.get(`core/${type}`) as Y.Array<unknown>

  // Create new deliverable from template
  const deliverable = Block.create({
    type: `core/${type}`,
    uuid: id,
    rel: 'deliverable',
    title: slug
  })

  // Group deliverable
  const [groupedDeliverable] = group([deliverable], 'type')[`core/${type}`]

  // Convert to YMap
  const yDeliverable = toYMap(
    groupedDeliverable as unknown as Record<string, unknown>,
    new Y.Map()
  )
  // Push to existing articles
  yDeliverables.push([yDeliverable])
}
