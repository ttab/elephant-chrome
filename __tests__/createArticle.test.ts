import { createArticle } from '@/views/Wire/lib/createArticle'
import { createDocument } from '@/lib/createYItem'
import * as Templates from '@/defaults/templates'
import type { Session } from 'next-auth'
import { vi } from 'vitest'
import type { HocuspocusProvider } from '@hocuspocus/provider'
import { fromYjsNewsDoc } from '../src-srv/utils/transformations/yjsNewsDoc'
import { fromGroupedNewsDoc } from '../src-srv/utils/transformations/groupedNewsDoc'
import type { Block, Document } from '@ttab/elephant-api/newsdoc'

describe('createArticle', () => {
  const sendStateless = vi.fn()
  const session = { accessToken: 'abc123' } as Session
  const provider = {
    sendStateless
  } as unknown as HocuspocusProvider

  let article: { document: Document }
  let planning: { document: Document }

  // Create reproducible UUIDs
  const mockRandomUUID = vi.spyOn(crypto, 'randomUUID')
  mockRandomUUID.mockImplementation(() => {
    const callCount = mockRandomUUID.mock.calls.length
    return `uuid-${callCount}`
  })

  // These are created in DialogView in the _real_ world
  beforeAll(() => {
    const originalPlanning = createDocument({
      template: Templates.planning,
      payload: { newsvalue: '4' }
    })
    const originalArticle = createDocument({
      template: Templates.article
    })

    const result = createArticle({
      documentId: originalArticle[0],
      document: originalArticle[1],
      title: '',
      provider,
      status: 'authenticated',
      session,
      planningDocument: originalPlanning[1],
      planningId: originalPlanning[0]
    })

    if (!result || !result.planning) {
      throw new Error('No planning document created')
    }

    if (!result || !result.planning) {
      throw new Error('No planning document created')
    }

    const planningYjs = fromYjsNewsDoc(result.planning).documentResponse
    const articleYjs = fromYjsNewsDoc(result.article).documentResponse
    article = fromGroupedNewsDoc(articleYjs)
    planning = fromGroupedNewsDoc(planningYjs)
  })

  it('should match planning snapshot', () => {
    expect(planning).toMatchSnapshot()
  })


  it('should match article snapshot', () => {
    expect(planning).toMatchSnapshot()
  })

  it('should have the same newsvalue in both planning and article', () => {
    const articleNewsvalue = article.document.meta
      .find((m: Block) => m.type === 'core/newsvalue' && m.value === '4')
    const planningNewsvalue = planning.document.meta
      .find((m: Block) => m.type === 'core/newsvalue' && m.value === '4')

    expect(articleNewsvalue).toEqual(planningNewsvalue)
  })


  it('should call sendStateless twice', () => {
    expect(sendStateless).toHaveBeenCalledTimes(2)
  })
})
