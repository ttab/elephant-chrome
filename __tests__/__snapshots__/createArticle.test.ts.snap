// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createArticle > should match article snapshot 1`] = `
{
  "document": {
    "content": [],
    "language": "sv-se",
    "links": [],
    "meta": [
      {
        "content": [],
        "contenttype": "",
        "data": {
          "end_date": "2025-02-14",
          "public": "true",
          "start_date": "2025-02-14",
          "tentative": "false",
        },
        "id": "",
        "links": [],
        "meta": [],
        "name": "",
        "rel": "",
        "role": "",
        "sensitivity": "",
        "title": "",
        "type": "core/planning-item",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "",
      },
      {
        "content": [],
        "contenttype": "",
        "data": {},
        "id": "",
        "links": [],
        "meta": [],
        "name": "",
        "rel": "",
        "role": "",
        "sensitivity": "",
        "title": "",
        "type": "core/newsvalue",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "4",
      },
      {
        "content": [],
        "contenttype": "",
        "data": {
          "end_date": "2025-02-14",
          "full_day": "true",
          "public": "true",
          "publish": "2024-02-09T10:30:00Z",
          "start": "2025-02-13T23:00:00.000Z",
          "start_date": "2025-02-14",
        },
        "id": "uuid-18",
        "links": [
          {
            "content": [],
            "contenttype": "",
            "data": {},
            "id": "",
            "links": [],
            "meta": [],
            "name": "",
            "rel": "deliverable",
            "role": "",
            "sensitivity": "",
            "title": "",
            "type": "core/article",
            "uri": "",
            "url": "",
            "uuid": "uuid-11",
            "value": "",
          },
        ],
        "meta": [
          {
            "content": [],
            "contenttype": "",
            "data": {},
            "id": "",
            "links": [],
            "meta": [],
            "name": "",
            "rel": "",
            "role": "",
            "sensitivity": "",
            "title": "",
            "type": "core/assignment-type",
            "uri": "",
            "url": "",
            "uuid": "",
            "value": "text",
          },
        ],
        "name": "",
        "rel": "",
        "role": "",
        "sensitivity": "",
        "title": "",
        "type": "core/assignment",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "",
      },
    ],
    "title": "",
    "type": "core/planning-item",
    "uri": "core://newscoverage/uuid-1",
    "url": "",
    "uuid": "uuid-1",
  },
  "isMetaDocument": false,
  "mainDocument": "",
  "version": 0n,
}
`;

exports[`createArticle > should match planning snapshot 1`] = `
{
  "document": {
    "content": [],
    "language": "sv-se",
    "links": [],
    "meta": [
      {
        "content": [],
        "contenttype": "",
        "data": {
          "end_date": "2025-02-14",
          "public": "true",
          "start_date": "2025-02-14",
          "tentative": "false",
        },
        "id": "",
        "links": [],
        "meta": [],
        "name": "",
        "rel": "",
        "role": "",
        "sensitivity": "",
        "title": "",
        "type": "core/planning-item",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "",
      },
      {
        "content": [],
        "contenttype": "",
        "data": {},
        "id": "",
        "links": [],
        "meta": [],
        "name": "",
        "rel": "",
        "role": "",
        "sensitivity": "",
        "title": "",
        "type": "core/newsvalue",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "4",
      },
      {
        "content": [],
        "contenttype": "",
        "data": {
          "end_date": "2025-02-14",
          "full_day": "true",
          "public": "true",
          "publish": "2024-02-09T10:30:00Z",
          "start": "2025-02-13T23:00:00.000Z",
          "start_date": "2025-02-14",
        },
        "id": "uuid-18",
        "links": [
          {
            "content": [],
            "contenttype": "",
            "data": {},
            "id": "",
            "links": [],
            "meta": [],
            "name": "",
            "rel": "deliverable",
            "role": "",
            "sensitivity": "",
            "title": "",
            "type": "core/article",
            "uri": "",
            "url": "",
            "uuid": "uuid-11",
            "value": "",
          },
        ],
        "meta": [
          {
            "content": [],
            "contenttype": "",
            "data": {},
            "id": "",
            "links": [],
            "meta": [],
            "name": "",
            "rel": "",
            "role": "",
            "sensitivity": "",
            "title": "",
            "type": "core/assignment-type",
            "uri": "",
            "url": "",
            "uuid": "",
            "value": "text",
          },
        ],
        "name": "",
        "rel": "",
        "role": "",
        "sensitivity": "",
        "title": "",
        "type": "core/assignment",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "",
      },
    ],
    "title": "",
    "type": "core/planning-item",
    "uri": "core://newscoverage/uuid-1",
    "url": "",
    "uuid": "uuid-1",
  },
  "isMetaDocument": false,
  "mainDocument": "",
  "version": 0n,
}
`;
